numhask/src/NumHask/Data/LogField.hs

```diff
@@ -0,0 +1,500 @@
+            )
+
+
+-- TODO: never worked with FFI, so i don't really know what's going on
+----------------------------------------------------------------
+-- Technically these should use 'Foreign.C.CDouble' however there's
+-- no isomorphism provided to normal 'Double'. The former is
+-- documented as being a newtype of the later, and so this should
+-- be safe.
+
+-- #ifdef __USE_FFI__
+-- #define LOG1P_WHICH_VERSION FFI version.
+-- #else
+-- #define LOG1P_WHICH_VERSION naive version! \
+--     Contact the maintainer with any FFI difficulties.
+-- #endif
+
+
+-- | Definition: @log1p == log . (1+)@. Standard C libraries provide
+-- a special definition for 'log1p' which is more accurate than
+-- doing the naive thing, especially for very small arguments. For
+-- example, the naive version underflows around @2 ** -53@, whereas
+-- the specialized version underflows around @2 ** -1074@. This
+-- function is used by ('+') and ('-') on @LogField@.
+--
+-- N.B. The @statistics:Statistics.Math@ module provides a pure
+-- Haskell implementation of @log1p@ for those who are interested.
+-- We do not copy it here because it relies on the @vector@ package
+-- which is non-portable. If there is sufficient interest, a portable
+-- variant of that implementation could be made. Contact the
+-- maintainer if the FFI and naive implementations are insufficient
+-- for your needs.
+--
+-- /This installation was compiled to use the LOG1P_WHICH_VERSION/
+
+-- #ifdef __USE_FFI__
+-- foreign import ccall unsafe "math.h log1p"
+--     log1p :: Double -> Double
+-- #else
+-- See statistics:Statistics.Math for a more accurate Haskell
+-- implementation.
+log1p :: ExpField a => a -> a
+{-# INLINE [0] log1p #-}
+log1p x = log (one + x)
+-- #endif
+
+
+-- | Definition: @expm1 == subtract 1 . exp@. Standard C libraries
+-- provide a special definition for 'expm1' which is more accurate
+-- than doing the naive thing, especially for very small arguments.
+-- This function isn't needed internally, but is provided for
+-- symmetry with 'log1p'.
+--
+-- /This installation was compiled to use the LOG1P_WHICH_VERSION/
+
+-- #ifdef __USE_FFI__
+-- foreign import ccall unsafe "math.h expm1"
+--     expm1 :: Double -> Double
+-- #else
+expm1 :: ExpField a => a -> a
+{-# INLINE [0] expm1 #-}
+expm1 x = exp x - one
+-- #endif
+
+-- CPP guarded because they won't fire if we're using the FFI versions
+-- #if !defined(__USE_FFI__)
+{-# RULES
+-- Into log-domain and back out
+"expm1/log1p"    forall x. expm1 (log1p x) = x
+
+-- Out of log-domain and back in
+"log1p/expm1"    forall x. log1p (expm1 x) = x
+    #-}
+-- #endif
+
+instance (ExpField a, LowerBoundedField a, Ord a) => AdditiveMagma (LogField a) where
+      (LogField x) `plus` (LogField y)
+            -- TODO unsure what to do here...currentl only negInfinity because of the 0-element
+            | x == y
+            && x == negInfinity
+            && y == negInfinity = LogField x -- @0+0 == 0@,
```

Comment:
= zero is fine for the first guard.  You could also test if either is zero.

```
x@(LogField x') `plus` y@(LogField y')
  | x == zero && y == zero = zero
  | x == zero = y
  | y == zero = x etc
```

Reply:
ah! yeah, that's smart 😉